### PR TITLE
Harden audit module review findings

### PR DESCRIPTION
**Pull request by rmusser01**
## Summary

- What changed:
bugfix audit 

---
## Summary by cubic
Hardened the audit module to resolve review findings (Linear TASK-2409): CSV exports neutralize spreadsheet formulas, reads flush buffers and propagate cancellation, and hash chaining and shared migration are reliable across instances. Updated README and added focused regression tests.

- **Bug Fixes**
  - CSV exports neutralize formula prefixes (=, +, -, @, tab, newline) across in-memory, streaming, and file writes.
  - PII redaction now covers metadata keys and preserves key uniqueness with stable suffixes.
  - query_events, count_events, export_events, and get_security_summary flush buffered events first; cancellations are re-raised instead of wrapped.
  - Hash-chain writes read the latest persisted head and run inside BEGIN IMMEDIATE transactions; replay writes follow the same pattern.
  - Shared migration resumes by rowid with fallback for rowid reuse and includes late older-timestamp rows.

<sup>Written for commit 41b67a1cbd4d39dcfdc30b5273ba8271865b1862. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rmusser01/tldw_server/pull/2440?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: tldw_Server_API/app/core/Audit/unified_audit_service.py [hunk 1]
```diff
@@ -3198,6 +3231,8 @@ async def count_events(
             async with self._read_db() as db, db.execute(query, params) as cursor:
                 row = await cursor.fetchone()
                 return int(row[0]) if row else 0
+        except asyncio.CancelledError:
+            raise
         except _AUDIT_NONCRITICAL_EXCEPTIONS as e:
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/unified_audit_service.py:3259]: Addressed. count_events already flushes with raise_on_failure before building the count query, and 41b67a1c also clarifies the cancellation test so the count path raises from the patched flush path rather than the earlier _read_db patch.

File: tldw_Server_API/tests/Audit/test_unified_audit_service.py [hunk 2]
```diff
@@ -2430,6 +2461,35 @@ async def _boom_read_db():
...
+    async def _cancelled_flush(*_args, **_kwargs):
+        raise asyncio.CancelledError()
+
+    monkeypatch.setattr(audit_service, "_read_db", _cancelled_read_db)
+    with pytest.raises(asyncio.CancelledError):
+        await audit_service.query_events(user_id="cancelled-query-user")
+
+    monkeypatch.setattr(audit_service, "flush", _cancelled_flush)
+    with pytest.raises(asyncio.CancelledError):
+        await audit_service.count_events(user_id="cancelled-count-user")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/Audit/test_unified_audit_service.py:2492]: Fixed in 41b67a1c. The test now stores the original _read_db and restores it before patching flush, so the count_events assertion verifies cancellation propagation from flush.

File: tldw_Server_API/app/core/Audit/unified_audit_service.py [hunk 3]
```diff
@@ -667,7 +678,11 @@ def redact_obj(self, data: Any, placeholder_format: str = "[{type}_REDACTED]") -
         """Recursively redact PII from nested dict/list/tuple/set structures and strings."""
         try:
             if isinstance(data, dict):
-                return {k: self.redact_obj(v, placeholder_format) for k, v in data.items()}
+                redacted: dict[Any, Any] = {}
+                for key, value in data.items():
+                    safe_key = self._redact_value(key, placeholder_format) if isinstance(key, str) else key
+                    redacted[safe_key] = self.redact_obj(value, placeholder_format)
+                return redacted
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/unified_audit_service.py:691]: Fixed in 21934984. Redacted metadata keys now retain colliding values with stable __N suffixes instead of overwriting earlier entries, and the PII redaction test covers this case.

File: tldw_Server_API/app/core/Audit/unified_audit_service.py [hunk 4]
```diff
@@ -220,6 +220,17 @@ def _normalize_storage_mode(raw: Any) -> str:
...
 
+def _neutralize_csv_formula(value: Any) -> Any:
+    """Prevent spreadsheet formula execution for exported string cells."""
+    if isinstance(value, str) and value and value[0] in _CSV_FORMULA_PREFIXES:
+        return "'" + value
+    return value
+
+
+def _neutralize_csv_row(row: dict[str, Any]) -> dict[str, Any]:
+    return {key: _neutralize_csv_formula(value) for key, value in row.items()}
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/unified_audit_service.py:232]: Fixed in 21934984. _neutralize_csv_row now has a docstring describing its CSV export sanitization role.

File: tldw_Server_API/tests/Audit/test_unified_audit_service.py [hunk 5]
```diff
@@ -2430,6 +2461,35 @@ async def _boom_read_db():
...
+    with pytest.raises(asyncio.CancelledError):
+        await audit_service.export_events(
+            format="json",
+            user_id="cancelled-export-user",
+            stream=False,
+            max_rows=10,
+        )
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/Audit/test_unified_audit_service.py:2499]: Fixed in 21934984. test_read_methods_propagate_cancellation now includes get_security_summary, so the summary read path preserves asyncio.CancelledError as expected.

File: tldw_Server_API/tests/Audit/test_audit_shared_migration.py [hunk 6]
```diff
@@ -256,6 +256,75 @@ async def test_migration_resume_checkpoint(tmp_path):
...
+        shared_db_path=shared_db_path,
+        user_db_base_dir=user_base,
+        default_db_path=None,
+        system_tenant_id="system",
+        chunk_size=100,
+    )
+    source_counts = next(c for c in report2.sources if c.source.label == f"user:{user_id}")
+    assert source_counts.events_read == 1
+    assert source_counts.events_inserted == 1
+
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/Audit/test_audit_shared_migration.py:319]: Fixed in 21934984. The late older timestamp migration test now runs a third pass and asserts events_read == 0 and events_inserted == 0 to lock in idempotent resume behavior.

File: tldw_Server_API/app/core/Audit/unified_audit_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/unified_audit_service.py:691]: Fixed in 21934984. Redacted metadata key collisions are assigned stable __N suffixes, preserving every original value, and the test asserts both colliding redacted keys remain present.

File: tldw_Server_API/tests/Audit/test_unified_audit_service.py [hunk 7]
```diff
@@ -2763,3 +2823,183 @@ async def test_export_events_flushes_buffered_events(tmp_path):
...
+    try:
+        await service.log_event(
+            AuditEventType.SECURITY_VIOLATION,
+            context=AuditContext(user_id="summary-user", ip_address="203.0.113.10"),
+            action="blocked_access",
+            result="failure",
+        )
+
+        summary = await service.get_security_summary(hours=1)
+
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/Audit/test_unified_audit_service.py:2888]: Fixed in 21934984. The security summary flush test disables risk scoring and asserts the event remains buffered before get_security_summary is called, so the test exercises the read-time flush path.

File: tldw_Server_API/app/core/Audit/audit_shared_migration.py [hunk 8]
```diff
@@ -743,15 +743,18 @@ async def _migrate_source(
...
                 query = (
                     "SELECT rowid, * FROM audit_events "
-                    "WHERE (timestamp > ? OR (timestamp = ? AND event_id > ?)) "
-                    "ORDER BY timestamp, event_id"
+                    "WHERE rowid > ? OR (timestamp > ? OR (timestamp = ? AND event_id > ?)) "
+                    "ORDER BY rowid"
                 )
-                params = (last_timestamp, last_timestamp, last_event_id or "")
+                params = (last_rowid, last_timestamp, last_timestamp, last_event_id or "")
+            elif last_rowid > 0:
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/audit_shared_migration.py]: Fixed in 21934984. Once a rowid checkpoint exists, resume selection is monotonic with WHERE rowid > ?, while legacy timestamp-only checkpoints keep their timestamp fallback path.

File: tldw_Server_API/app/core/Audit/audit_shared_migration.py [hunk 9]
```diff
@@ -743,15 +743,18 @@ async def _migrate_source(
...
+                query = "SELECT rowid, * FROM audit_events WHERE rowid > ? ORDER BY rowid"
+                params = (last_rowid,)
             else:
-                query = "SELECT rowid, * FROM audit_events ORDER BY timestamp, event_id"
+                query = "SELECT rowid, * FROM audit_events ORDER BY rowid"
                 params = ()
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/audit_shared_migration.py]: Fixed in 21934984. Normal migration resume now uses a rowid-only cursor once last_rowid is present; timestamp fallback is only used for legacy timestamp-only checkpoints. Rowid reuse/source rebuild cases reset to a full scan with event_id de-duplication.

File: tldw_Server_API/app/core/Audit/unified_audit_service.py [hunk 10]
```diff
@@ -121,6 +120,7 @@ def _current_app_settings() -> Any:
 _SYSTEM_TENANT_ID = "system"
 _UNIDENTIFIED_TENANT_ID = "unidentified_user"
 _AUDIT_SHARED_SCHEMA_VERSION = 1
+_CSV_FORMULA_PREFIXES = ("=", "+", "-", "@", "\t", "\r")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/unified_audit_service.py]: Fixed in 21934984. LF is included in _CSV_FORMULA_PREFIXES, and the CSV export test covers a newline-prefixed formula payload.

File: tldw_Server_API/app/core/Audit/audit_shared_migration.py [hunk 11]
```diff
@@ -743,15 +743,18 @@ async def _migrate_source(
...
+            if last_rowid > 0 and last_timestamp:
                 query = (
                     "SELECT rowid, * FROM audit_events "
-                    "WHERE (timestamp > ? OR (timestamp = ? AND event_id > ?)) "
-                    "ORDER BY timestamp, event_id"
+                    "WHERE rowid > ? OR (timestamp > ? OR (timestamp = ? AND event_id > ?)) "
+                    "ORDER BY rowid"
                 )
-                params = (last_timestamp, last_timestamp, last_event_id or "")
+                params = (last_rowid, last_timestamp, last_timestamp, last_event_id or "")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Audit/audit_shared_migration.py]: Fixed in 21934984. The migration resume query no longer mixes rowid and timestamp predicates for normal checkpoints; the added late older timestamp and rowid-reuse tests cover the resume edge cases.

File: tldw_Server_API/tests/Audit/test_unified_audit_service.py [hunk 12]
```diff
@@ -2763,3 +2823,183 @@ async def test_export_events_flushes_buffered_events(tmp_path):
...
+        output_path = tmp_path / "audit_export.csv"
+        written = await service.export_events(
+            user_id="csv-formula-user",
+            format="csv",
+            file_path=output_path,
+            chunk_size=1,
+            max_rows=10,
+        )
+        assert written == 1
+        rows = list(csv.DictReader(StringIO(output_path.read_text(encoding="utf-8"))))
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/Audit/test_unified_audit_service.py]: Fixed in 21934984. The async CSV export test now reads the file through asyncio.to_thread instead of calling Path.read_text directly on the event loop.
